pkg/yurttunnel/server/serveraddr/addr_test.go

```diff
@@ -50,3 +56,185 @@ func TestGetDefaultDomainsForSvc(t *testing.T) {
+			expectValue: ExpectValue{
+				ErrorMsg: "there is no cloud node",
+			},
+		},
+		{
+			desc: "there is no cloud node",
+			nodelist: v1.NodeList{
+				Items: []v1.Node{
+					{
+						Status: v1.NodeStatus{
+							Addresses: []v1.NodeAddress{
+								{
+									Type:    v1.NodeInternalDNS,
+									Address: "192.168.1.1",
+								},
+							},
+						},
+					},
+				},
+			},
+			expectValue: ExpectValue{
+				ErrorMsg: "can't find node IP",
+			},
+		},
+		{
+			desc: "Get IPs Error",
+			nodelist: v1.NodeList{
+				Items: []v1.Node{
+					{
+						Status: v1.NodeStatus{
+							Addresses: []v1.NodeAddress{
+								{
+									Type:    v1.NodeInternalIP,
+									Address: "192.168.1.1",
+								},
+							},
+						},
+					},
+				},
+			},
+			expectValue: ExpectValue{
+				ErrorMsg: "Get IPs Error",
+				ips: []net.IP{
+					net.ParseIP("192.168.1.1"),
+				},
+			},
+		},
+		{
+			desc: "Many IPs",
+			nodelist: v1.NodeList{
+				Items: []v1.Node{
+					{
+						Status: v1.NodeStatus{
+							Addresses: []v1.NodeAddress{
+								{
+									Type:    v1.NodeInternalIP,
+									Address: "192.168.1.1",
+								},
+								{
+									Type:    v1.NodeHostName,
+									Address: "192.168.1.2",
+								},
+								{
+									Type:    v1.NodeExternalDNS,
+									Address: "192.168.1.3",
+								},
+								{
+									Type:    v1.NodeInternalIP,
+									Address: "192.168.1.4",
+								},
+							},
+						},
+					},
+				},
+			},
+			expectValue: ExpectValue{
+				ErrorMsg: "IPs is Invalid.",
+				ips: []net.IP{
+					net.ParseIP("192.168.1.1"),
+					net.ParseIP("192.168.1.4"),
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			//getNodePortDNSandIP(nodeLst *v1.NodeList) ([]string, []net.IP, error)
+			_, getIPS, error := getNodePortDNSandIP(&tt.nodelist)
+			if error != nil {
+				if error.Error() != tt.expectValue.ErrorMsg {
+					t.Errorf("error is %s\n", tt.expectValue.ErrorMsg)
+				}
+			} else {
+				if !reflect.DeepEqual(getIPS, tt.expectValue.ips) {
+					t.Error(tt.expectValue.ErrorMsg)
+				}
+			}
+		})
+	}
+
+}
+
+func TestGetClusterIPDNSandIP(t *testing.T) {
+	type ExpectValue struct {
+		ErrorMsg string
+		dnsNames []string
+		ips      []net.IP
+	}
+
+	tests := []struct {
+		desc        string
+		svc         corev1.Service
+		expectValue ExpectValue
+	}{
+		{
+			desc: "there is constants.YurttunnelServerExternalAddrKey ips is  192.168.1.2",
+			svc: corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						constants.YurttunnelServerExternalAddrKey: "192.168.1.2:8080",
+					},
+				},
+			},
+			expectValue: ExpectValue{
+				ErrorMsg: "there is constants.YurttunnelServerExternalAddrKey ips is  192.168.1.2",
+				ips: []net.IP{
+					net.ParseIP("192.168.1.2"),
+				},
+				dnsNames: []string{},
+			},
+		},
+		{
+			desc: "there is constants.YurttunnelServerExternalAddrKey dnsnames is  www.bing.com:80",
```

Comment:
"constants.YurttunnelServerExternalAddrKey exists and is associated to dnsnames ["www.bing.com:80"]"

Reply:
I will modify it later

Reply:
I did it